### DISCUSSION TITLE
Concurrent transactions strange behaviour

### DISCUSSION THREAD

**Discussion by MrVhek** [prisma/prisma#23200] [Q&A] [answered]
### Question

Hi,
I have an issue with prisma usage : I have an entrypoint that can be called a certain number of time and sometimes with the same values due to different reasons. I have some constraint in my database that should make things fail for multiple inserts of the same data. I use postgres with default transaction isolation level.

Here my prisma logs : I have 2 transaction ongoing with the same data inserted. The issue is that the first transaction should succeed, while the second should failed. I put the first one in bold

**2024-02-20T14:33:31.132Z [igonogo API] debug: BEGIN: duration: 0 ms**

2024-02-20T14:33:31.146Z [igonogo API] debug: BEGIN: duration: 0 ms

**2024-02-20T14:33:31.156Z [igonogo API] debug: insert into "VariableValues" ("variableId", "timestamp", "type", "resultTable", "resultId", "participantUid") values ($1, $2, $3::"VariableTypes", $4::"ResultsTableTypes", $5, $6), ($7, $8, $9::"VariableTypes", $10::"ResultsTableTypes", $11, $12) returning "VariableValues"."uid": duration: 6 ms**

2024-02-20T14:33:31.182Z [igonogo API] debug: insert into "VariableValues" ("variableId", "timestamp", "type", "resultTable", "resultId", "participantUid") values ($1, $2, $3::"VariableTypes", $4::"ResultsTableTypes", $5, $6), ($7, $8, $9::"VariableTypes", $10::"ResultsTableTypes", $11, $12) returning "VariableValues"."uid": duration: 20 ms

2024-02-20T14:33:31.206Z [igonogo API] debug: ROLLBACK: duration: 10 ms

2024-02-20T14:33:31.209Z [igonogo API] error: PrismaClientKnownRequestError: 
Raw query failed. Code: `23505`. Message: `Key ("variableId", type, "timestamp", "participantUid")=(14745, BINARY, 2024-02-20 14:33:30.689, test) already exists.`

**2024-02-20T14:33:31.212Z [igonogo API] debug: INSERT INTO "public"."VariableFloatValues" ("value","type","variableValueUid") VALUES ($1,CAST($2::text AS "public"."VariableTypes"),$3): duration: 52 ms**

**2024-02-20T14:33:31.229Z [igonogo API] debug: ROLLBACK: duration: 7 ms
Foreign key constraint failed on the field: `variablefloatvalues_variable_value_uid_type_fk (index)`**

I expect the first transaction to succeed and the second one to fail because the value already exists. But the first transaction is also failing for an obscure reason.

Anyone can explain what is happening ?

Thanks !

### Expected behavior (optional)

I expect the first transaction to succeed while the second to fail, but here the two are failing

### Information about Prisma Schema, Client Queries and Environment (optional)

I have a table VariableFloatValue that have a foreign key to table VariableValue.

- OS: macOS
- Database: PostgreSQL
- Node.js version: v18.18.2
- Run `prisma -v` to see your Prisma version and paste it: 5.9.1

**Comment by nurul3101** [comment]
Hey @MrVhek 👋 

Your logs indicate that the transactions are started almost simultaneously and attempt to insert duplicate data. Due to the race condition, the second transaction isn't aware of the first one's changes until it tries to commit, resulting in both transactions failing. This is because, by the time the first transaction commits, the second one has already attempted its insert and is pending completion. When the unique constraint is finally checked, it fails for both.

To ensure that one transaction completes before another begins when they involve the same data, consider using the `Serializable` isolation level. This is the strictest level, where transactions are fully isolated from each other, essentially executing as if they were serialized. 

Here's how you can set the serializable isolation level in Prisma: [Reference](https://www.prisma.io/docs/orm/prisma-client/queries/transactions#set-the-isolation-level)

If this answers your question, it would be great if you could **mark this Discussion as answered** to indicate that it has been resolved. 

Otherwise please let us know how else we can help you further or close the Discussion if it was resolved in some other way 🙏

**Reply by MrVhek** [reply]
Hey @nurul3101 
Thanks a lot for your help. I'm not sure to understand everything in what you said, here the first transaction fail because of a foreign key not met, but it is actually inserted before in the same transaction, so I don't see why this transaction first insert has been rollbacked with the rollback of the second transaction. I expect that the second transaction rollback would not impact the first transaction, is this not how it works ?

Thanks a lot

**Comment by MrVhek** [comment, marked as answer]
This still doesn't work, I have a deadlock on my transaction now...(I added an unique constraint on the DragAndDrop table that should fail)

To give more context and testing:
I use the same prisma client with these options:
```
maxWait: 10000, // default: 2000
timeout: 20000, // default: 5000
isolationLevel: Prisma.TransactionIsolationLevel.Serializable,
```

 I have an unit test that is inserting data in multiple table twice like this:
```
 console.log("-------------------")
const [res, res2] = await Promise.all([
prisma
	.$transaction(async (tx: PrismaTransaction) => {
		return await addDragAndDropResultV1Bugged(
			validDragAndDropResult,
			[0, 0, 500, 200],
			tx,
			"firstTransaction",
		)
	})
	.then((res) => ({...res, status: 201}))
	.catch((e) => {
		console.error("Now error")
		return {status: 500, errror: e.message}
	}),
prisma
	.$transaction(async (tx: PrismaTransaction) => {
		return await addDragAndDropResultV1Bugged(
			validDragAndDropResult,
			[200, 0, 0, 0],
			tx,
			"secondTransaction",
		)
	})
	.then((res) => ({...res, status: 201}))
	.catch((e) => {
		console.error("Now error")
		return {status: 500, errror: e.message}
	}),
])
console.log("-------------------")
```
Here the logs:
https://pastebin.com/raw/ZZveZtee

The order of the function is : Create DragAndDropResult -> Select Variables -> Create Variable Values -> Create Variable Value Float -> Create Variable Value Float

EDIT: finally, the issue is with a tier package that I used named prisma-extension-kysely
